### ISSUE TITLE
Relu different speed - optimization not working?

### ISSUE THREAD

**Issue by albertz** [Theano/Theano#2698]
Relu is a quite simple function and I would expect that different definitions would all be optimized to the fastest calculation and thus that all different definitions would have the same speed. However, that seems not the case.

1) `T.switch(x < 0., 0., x)`
2) `x * (x > 0)`
3) `T.maximum(0.0, x)`
4) `(T.sgn(x) + 1) * x * 0.5`

From [here](https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/scripts/benchmark/time_relu.py) or [here](http://stackoverflow.com/questions/26497564/theano-hiddenlayer-activation-function), it seems as if `T.maximum` is really the slowest on CPU for both fprop and bprop. On GPU, they are similar in fprop but not in bprop where `T.maximum` is the fastest.

**Comment by f0k** [comment]
In Lasagne, we found the last formulation to give the best results on GPU both for the forward and backward pass, because it gets compiled into a single GpuElemwise Op: https://github.com/benanne/Lasagne/pull/163#issuecomment-81806482

Theano can only optimize expressions it has an explicit optimization rule for. While it would be easy to introduce an optimization that replaces `T.maximum(0, x)` with `0.5 * (x + abs(x))`, that wouldn't help for the backward pass -- `theano.grad()` would still be called on the original expression, so it would need another optimization to replace the expression for the gradient of `T.maximum(0, x)` with the expression for the gradient of `0.5 * (x + abs(x))`. For leaky rectified linear units, things would get even more complicated.

An automatic optimization of this would be great, but for now it's probably easier to use a more efficient formulation manually if you're running experiments on a GPU.

By the way, does anybody know if cuDNN would be faster for this? I've had instances where the rectified nonlinearity took about the same time as the first convolution in a network, so it could be worth optimizing.

**Comment by nouiz** [comment]
We could add in Theano a relu method that use the fastest implementation.
That way people do not need to benchmark it. Do you want to push it to
Theano?

If you find case where relu is significatively slow, tell us. We need to
know the input shape and strides. This could be the reason. I didn't plan
to use cudnn for that type of stuff as I do not expect it to give
significant speed up, but if you find place where it is much slower then
expected, this should be checked.

**Comment by gforge** [comment]
Not sure how active this issue is but I played around with the ReLU implementations and they are mostly the same, although the abs() variant is the fastest. I've also implemented the PReLU that was recently popularised and the same abs()-variant performed the best, I've written a blog post about it here: http://gforge.se/2015/06/benchmarking-relu-and-prelu/

Being new to the field I think it would be useful to have pre-optimized ReLU and PReLU functions. Since the activators are at the core of deep learning it seems like a reasonable thing to make sure that everyone uses the optimal version.

**Comment by nouiz** [comment]
That would be great. Do you want to make a pr with this?

Fred

**Comment by gforge** [comment]
A pull request? Sure, although I'm uncertain if there is a better way to define the derivatives. The general grad() seems to work great, but in some cases it could possibly be beneficial to provide a simpler alternative. For instance it seems that the fill() relies on the original size and needs the original calculation for it to work - if the forward propagation is used here there is no issue but if there is another calculation being performed here I would imagine it being a time consumer.

**Comment by f0k** [comment]
I've replied to your blog post. To the best of my knowledge (see the github issue linked further above), the fastest option for GPUs is this one:
```python
def rectify(x, alpha=0):
    f1 = 0.5 * (1 + alpha)
    f2 = 0.5 * (1 - alpha)
    return f1 * x + f2 * abs(x)
```
This is what we have in Lasagne. I could add that as `theano.tensor.nnet.rectify` or `theano.tensor.nnet.relu`, possibly with a special case for `alpha == 0` (that would currently result in `0.5 * x + 0.5 * abs(x)`, but `0.5 * (x + abs(x))` might be simpler).

As @gforge said, we might instead want to define an Op with a faster gradient implementation, though. Something like `T.switch(out > 0, 1, alpha)` will only work for positive `alpha`, but that might still speed up two common cases (the standard rectifier and the leaky rectifier with fixed leakiness), and for PReLU it could fall back to the standard `grad()` of the forward expression.
For reference, the standard gradient expression of the function above is (for fixed alpha):
```
>>> x = theano.tensor.matrix('x')
>>> y = rectify(x, alpha=0.3)
>>> out = theano.tensor.matrix('out')
>>> g = theano.grad(None, x, known_grads={y: out})
>>> f = theano.function([x, out], g)
>>> theano.printing.pp(f2.maker.fgraph.outputs[0])
'HostFromGpu(GpuElemwise{Composite{((i0 * i1) + (i2 * i1 * sgn(i3)))}}[(0, 1)](CudaNdarrayConstant{[[ 0.64999998]]}, GpuFromHost(out), CudaNdarrayConstant{[[ 0.34999999]]}, GpuFromHost(x)))'
```
That's probably difficult to top, given that it gets compiled into a single `GpuElemwise` (that's also why it's the fastest option of the ones I benchmarked). `theano.tensor.switch(out > 0, 1, 0.3)` also gets compiled into a single `GpuElemwise` that doesn't depend on `x`, but simple element-wise Ops seem to perform about the same -- and as I said above, this expression only works for positive alpha. (Plus I haven't thought through if it would do the correct thing if alpha is a vector or matrix.)

In summary, I'm not sure an Op would provide much of a benefit over the simple `rectify()` function I proposed above. Any more opinions? Shall I just file a PR adding the function along with tests?

**Comment by lamblin** [comment]
An option would be to have an Op computing the forward ReLU, an Op for the gradient, and a couple of optimizations to replace it with whatever implementation is more efficient on the current device.
For instance, the final expression for CPU and GPU may be different, and the expression for the gradient may be different from what `theano.grad` would have returned on the forward expression.

**Comment by nouiz** [comment]
I agree that would be the ideal solution, but I think it is overkill. I
never saw relu being a bottleneck.

So a quick good solution I think is to make one relu function in Theano
with a suggested version. We could use the flag device to swap a CPU vs GPU
implementation if you want to be more fency, but I would not even bother
with that before I see in a real model that relu take a significant % of
time.

**Comment by f0k** [comment]
An Op would have the additional benefit that you could add an optimizer to merge it with Nervana's convolution. But I agree that it's probably overkill for now. We can add a `theano.tensor.nnet.relu` function using the code I posted above and still have it insert an Op later if we think it's worth it, or at least add a `if theano.config.device == 'cpu'` switch with the fastest CPU option (still to be determined).

**Comment by nouiz** [comment]
Do you want to make a PR with your implementation?

**Comment by f0k** [comment]
> Do you want to make a PR with your implementation?

Can do. Where do I put it? `tensor/nnet/nnet.py` is a bloody mess of softmax- and cross-entropy related Ops and optimizers, shall I put it at the end of that or create a new file? All other existing files in `tensor/nnet/` are not a good fit either.

**Comment by nouiz** [comment]
I let you choose. You can make a new file. But a new file for what 10
lines, not sure it is worth it. Anyway, where ever you put it, import the
relu function in nnet/__init__.py. That way,  the user don't care if we
refactor later.

**Comment by f0k** [comment]
> I let you choose. You can make a new file. But a new file for what 10 lines, not sure it is worth it.

All right, I'll put it in `nnet.py` and we can pull it out if we ever decide to make an Op with optimizers out of it.

**Comment by chrwolf** [comment]
Hey, everyone. I played around with the ReLU activation as part of a larger model and found, that in the case, where both relu(x) and grad(relu(x)) appear in the function graph (often happens if gradients are taken), the T.maximum(x) implementation can be faster than the 0.5*(x + abs(x)) form. The reason being, that grad(T.maximum(x), x) is EQ(T.maximum(x), x), so the computations can be reused, whereas the gradient of the abs()-version doesn't allow any reuse.
However the gradient is 0.5 + 0.5*sgn(x), which can be rewritten as EQ(abs(x), x), which is not only much faster (abs seems to be much better than sgn), but also allows reusing computation results, if both relu(x) and grad(relu(x)) appear.
I implemented this by writing a PatternSub:
<code>local_relu_speedup = gof.PatternSub(in_pattern=(tensor.add, 0.5, (tensor.mul, 0.5, (tensor.sgn, 'x'))),
                                    out_pattern=(tensor.mul, 1.0, (tensor.eq, 'x', (tensor.abs_, 'x'))),
                                    name='Grad of ReLU optimization',
                                    allow_multiple_clients=True)
opt.register_canonicalize(local_relu_speedup, 'fast_run', name='Grad of ReLU optimization')</code>
To me this appears more like a temporary fix, since this depends on other optimizers not changing etc., but my attempts at creating the Op and making that work as fast, so this is all I've got at the moment.
Hope this helps :)

P.S.: For the case alpha unequal 0, I haven't tested anything, but the corresponding gradient should probably be: alpha + (1-alpha)*EQ(x, abs(x))
Judging from the alpha=0 case, this computation should be fast and computations can be merged

**Comment by nouiz** [comment]
The problem is that the answer isn,t the same for the CPU or GPU code.

You could register this later, when the computation was moved to the GPU,
so this would be done only for CPU.

For this register it in register_specialize_device.

But i'm not sure if this will work without change. I think Elemwise fusion
will have been applied already.

To allow this, we need to register an optimization only if the "gpu"
optimization tag isn't used. There isn't such a mechanism right now.

Before digging in this, how much speed up do you see in a real model? I was
under the impression that this don't make a big difference on full model.

**Comment by chrwolf** [comment]
Sorry, for the late response, just ran some more tests on this. All of these results are from computations using the GPU.
My model with the theano ReLU implementation takes approx. 8.1 sec per mini-batch gradient calculation.
With my implementation, which also uses Abs in the gradient, it takes (same machine, successive experiments with otherwise identical setup) approx. 6.3 sec per mini-batch gradient.
So in my case it makes a significant difference.
Below I added the profiles of the core scan node, where most of the work happens. It appears that the difference is due to different optimizations being applied, leading to different operations being called. To avoid problems with the PatternSub I used above I wrote a ReLU op using Abs in the gradient and then replaced it via PatternSub with 0.5*(x + Abs(x)) (really not sure how to otherwise do it as an Op), see the Code after the profiles.

Theano ReLU:
```
Scan Op profiling ( grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn )
==================
  Message: None
  Time in 30 calls of the op (for a total of 210 steps) 6.954700e+01s

  Total time spent in calling the VM 6.617300e+01s (95.149%)
  Total overhead (computing slices..) 3.373999e+00s (4.851%)

Time in all call to theano.grad() 1.299850e+02s
Time since theano import 1402.196s
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  22.9%    22.9%      15.085s       2.38e-04s     C    63420     302   theano.sandbox.cuda.basic_ops.GpuElemwise
  19.3%    42.2%      12.698s       3.36e-04s     C    37800     180   theano.sandbox.cuda.basic_ops.GpuIncSubtensor
  17.9%    60.1%      11.825s       3.78e-04s     C    31290     149   theano.sandbox.cuda.blas.GpuDot22
  17.4%    77.5%      11.435s       3.73e-04s     C    30660     146   theano.sandbox.cuda.blas.GpuDot22Scalar
  12.3%    89.8%       8.127s       3.99e-04s     C    20370      97   theano.sandbox.cuda.blas.GpuGemm
   9.9%    99.7%       6.536s       2.59e-04s     C    25200     120   theano.sandbox.cuda.basic_ops.GpuCAReduce
   0.1%    99.9%       0.094s       2.48e-06s     C    37800     180   theano.sandbox.cuda.basic_ops.GpuReshape
   0.1%   100.0%       0.062s       2.36e-06s     C    26460     126   theano.sandbox.cuda.basic_ops.GpuDimShuffle
   0.0%   100.0%       0.031s       1.49e-04s     Py     210       1   theano.compile.ops.SpecifyShape
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  17.9%    17.9%      11.825s       3.78e-04s     C     31290      149   GpuDot22
  17.4%    35.3%      11.435s       3.73e-04s     C     30660      146   GpuDot22Scalar
  15.1%    50.4%       9.953s       2.82e-04s     C     35280      168   GpuIncSubtensor{InplaceInc;int64:int64:}
  10.3%    60.7%       6.755s       3.83e-04s     C     17640       84   GpuGemm{inplace}
   7.3%    68.0%       4.820s       2.73e-04s     C     17640       84   GpuCAReduce{add}{1,0}
   4.9%    72.8%       3.198s       2.24e-04s     C     14280       68   GpuElemwise{mul,no_inplace}
   4.2%    77.0%       2.746s       1.09e-03s     C     2520       12   GpuIncSubtensor{Inc;int64:int64:}
   3.3%    80.3%       2.184s       2.31e-04s     C     9450       45   GpuElemwise{Composite{(((i0 + (i0 * i1)) + ((i2 * i3) / i4)) + ((i5 * i3 * i1 * i1) / i4))}}[(0, 0)]
   2.8%    83.1%       1.856s       2.60e-04s     C     7140       34   GpuElemwise{Composite{(i0 * (i1 + (i1 * i2)))}}[(0, 1)]
   2.6%    85.7%       1.716s       2.27e-04s     C     7560       36   GpuCAReduce{add}{1,1}
   2.1%    87.9%       1.404s       2.91e-04s     C     4830       23   GpuElemwise{Mul}[(0, 0)]
   2.1%    89.9%       1.373s       5.03e-04s     C     2730       13   GpuGemm{no_inplace}
   1.2%    91.1%       0.796s       2.91e-04s     C     2730       13   GpuElemwise{add,no_inplace}
   1.2%    92.3%       0.764s       2.27e-04s     C     3360       16   GpuElemwise{Mul}[(0, 1)]
   1.1%    93.4%       0.749s       2.55e-04s     C     2940       14   GpuElemwise{Composite{(i0 * (i1 + (i1 * i2)))},no_inplace}
   1.0%    94.4%       0.655s       2.60e-04s     C     2520       12   GpuElemwise{Composite{((i0 * i1 * i2) / i3)},no_inplace}
   0.9%    95.3%       0.593s       2.35e-04s     C     2520       12   GpuElemwise{Composite{((i0 * i1 * i2) + i3 + i4 + i5)},no_inplace}
   0.9%    96.2%       0.577s       2.11e-04s     C     2730       13   GpuElemwise{Mul}[(0, 2)]
   0.8%    97.0%       0.530s       2.10e-04s     C     2520       12   GpuElemwise{Composite{((i0 + i1) + i2)},no_inplace}
   0.8%    97.8%       0.530s       2.10e-04s     C     2520       12   GpuElemwise{Composite{((i0 * i1) * i2)},no_inplace}
   ... (remaining 20 Ops account for   2.18%(1.44s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
   0.4%     0.4%       0.281s       1.34e-03s    210   1294   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     0.9%       0.281s       1.34e-03s    210   1218   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     1.3%       0.265s       1.26e-03s    210   1293   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     1.6%       0.250s       1.19e-03s    210   1297   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     2.0%       0.234s       1.11e-03s    210   1298   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     2.3%       0.218s       1.04e-03s    210   1217   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     2.7%       0.218s       1.04e-03s    210   1292   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     3.0%       0.218s       1.04e-03s    210   1300   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     3.3%       0.218s       1.04e-03s    210   397   GpuDot22Scalar(GpuElemwise{Mul}[(0, 0)].0, <CudaNdarrayType(float32, matrix)>, TensorConstant{0.5})
   0.3%     3.6%       0.218s       1.04e-03s    210   1296   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     4.0%       0.203s       9.66e-04s    210   1031   GpuDot22Scalar(GpuElemwise{Composite{(((i0 + (i0 * i1)) + ((i2 * i3) / i4)) + ((i5 * i3 * i1 * i1) / i4))}}[(0, 0)].0, <CudaNdarrayType(float32, matrix)>, TensorConstant{0.5})
   0.3%     4.3%       0.203s       9.66e-04s    210   1219   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     4.6%       0.203s       9.66e-04s    210   402   GpuDot22(GpuDimShuffle{1,0}.0, GpuElemwise{Mul}[(0, 0)].0)
   0.3%     4.9%       0.187s       8.91e-04s    210   1295   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     5.1%       0.187s       8.91e-04s    210   991   GpuGemm{inplace}(GpuDot22.0, TensorConstant{1.0}, GpuDimShuffle{1,0}.0, GpuElemwise{Composite{(i0 + (-i1))},no_inplace}.0, TensorConstant{1.0})
   0.3%     5.4%       0.187s       8.91e-04s    210   424   GpuGemm{inplace}(GpuDot22.0, TensorConstant{1.0}, GpuDimShuffle{1,0}.0, GpuElemwise{Composite{(i0 + (-i1))},no_inplace}.0, TensorConstant{1.0})
   0.3%     5.7%       0.187s       8.91e-04s    210   263   GpuDot22Scalar(GpuElemwise{Composite{(i0 * (i1 + (i1 * i2)))}}[(0, 1)].0, <CudaNdarrayType(float32, matrix)>, TensorConstant{0.5})
   0.3%     6.0%       0.187s       8.91e-04s    210   451   GpuGemm{inplace}(GpuDot22.0, TensorConstant{1.0}, GpuDimShuffle{1,0}.0, GpuElemwise{Composite{(i0 + (-i1))},no_inplace}.0, TensorConstant{1.0})
   0.3%     6.2%       0.172s       8.17e-04s    210   1299   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     6.5%       0.172s       8.17e-04s    210   1091   GpuGemm{no_inplace}(<CudaNdarrayType(float32, matrix)>, TensorConstant{1.0}, GpuElemwise{Composite{(((i0 + (i0 * i1)) + ((i2 * i3) / i4)) + ((i5 * i3 * i1 * i1) / i4))}}[(0, 0)].0, <CudaNdarrayType(float32, matrix)>, TensorConstant{1.0})
   ... (remaining 1281 Apply instances account for 93.49%(61.60s) of the runtime)
   
```   
   
Gradient with Abs implementation:
```   
 Scan Op profiling ( grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn&grad_of_scan_fn )
==================
  Message: None
  Time in 30 calls of the op (for a total of 210 steps) 5.384500e+01s

  Total time spent in calling the VM 5.100800e+01s (94.731%)
  Total overhead (computing slices..) 2.836998e+00s (5.269%)

Time in all call to theano.grad() 1.093870e+02s
Time since theano import 726.676s
Class
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Class name>
  28.6%    28.6%      14.507s       3.51e-04s     C    41370     197   theano.sandbox.cuda.blas.GpuDot22
  24.9%    53.4%      12.636s       1.99e-04s     C    63420     302   theano.sandbox.cuda.basic_ops.GpuElemwise
  20.7%    74.1%      10.530s       2.79e-04s     C    37800     180   theano.sandbox.cuda.basic_ops.GpuIncSubtensor
  13.3%    87.4%       6.755s       3.32e-04s     C    20370      97   theano.sandbox.cuda.blas.GpuGemm
  12.3%    99.7%       6.240s       2.48e-04s     C    25200     120   theano.sandbox.cuda.basic_ops.GpuCAReduce
   0.2%   100.0%       0.125s       3.30e-06s     C    37800     180   theano.sandbox.cuda.basic_ops.GpuReshape
   0.0%   100.0%       0.016s       5.90e-07s     C    26460     126   theano.sandbox.cuda.basic_ops.GpuDimShuffle
   0.0%   100.0%       0.000s       0.00e+00s     Py     210       1   theano.compile.ops.SpecifyShape
   ... (remaining 0 Classes account for   0.00%(0.00s) of the runtime)

Ops
---
<% time> <sum %> <apply time> <time per call> <type> <#call> <#apply> <Op name>
  28.6%    28.6%      14.507s       3.51e-04s     C     41370      197   GpuDot22
  15.7%    44.2%       7.971s       2.26e-04s     C     35280      168   GpuIncSubtensor{InplaceInc;int64:int64:}
  11.2%    55.4%       5.694s       3.23e-04s     C     17640       84   GpuGemm{inplace}
   9.6%    65.1%       4.898s       2.78e-04s     C     17640       84   GpuCAReduce{add}{1,0}
   8.4%    73.4%       4.243s       1.98e-04s     C     21420      102   GpuElemwise{Mul}[(0, 0)]
   7.7%    81.1%       3.915s       2.00e-04s     C     19530       93   GpuElemwise{mul,no_inplace}
   5.0%    86.2%       2.558s       1.02e-03s     C     2520       12   GpuIncSubtensor{Inc;int64:int64:}
   2.6%    88.8%       1.342s       1.77e-04s     C     7560       36   GpuCAReduce{add}{1,1}
   2.1%    90.9%       1.061s       3.89e-04s     C     2730       13   GpuGemm{no_inplace}
   1.3%    92.2%       0.671s       2.46e-04s     C     2730       13   GpuElemwise{add,no_inplace}
   1.2%    93.4%       0.608s       2.41e-04s     C     2520       12   GpuElemwise{Composite{((i0 * i1 * i2) / i3)},no_inplace}
   1.0%    94.5%       0.530s       1.94e-04s     C     2730       13   GpuElemwise{Mul}[(0, 2)]
   1.0%    95.5%       0.530s       2.10e-04s     C     2520       12   GpuElemwise{Composite{((i0 + i1) + i2)},no_inplace}
   1.0%    96.5%       0.484s       1.92e-04s     C     2520       12   GpuElemwise{Composite{((i0 * i1 * i2) + i3 + i4 + i5)},no_inplace}
   1.0%    97.4%       0.484s       1.92e-04s     C     2520       12   GpuElemwise{Composite{((i0 * i1) * i2)},no_inplace}
   0.7%    98.2%       0.374s       1.78e-04s     C     2100       10   GpuElemwise{Mul}[(0, 1)]
   0.7%    98.9%       0.359s       1.42e-04s     C     2520       12   GpuElemwise{Composite{(i0 * (i1 + (-i2)) * i3)}}[(0, 1)]
   0.2%    99.1%       0.125s       3.30e-06s     C     37800      180   GpuReshape{1}
   0.2%    99.3%       0.078s       1.86e-04s     C      420        2   GpuElemwise{Composite{(i0 * (i1 + i2))},no_inplace}
   0.2%    99.4%       0.078s       3.71e-04s     C      210        1   GpuElemwise{Composite{(Composite{(i0 + (i0 + (i0 + (i0 + (i0 + (i0 + i0))))))}(i0) + (i0 + Composite{(i0 + (i0 + (i0 + (i0 + (i0 + (i0 + i0))))))}(i0)))},no_inplace}
   ... (remaining 13 Ops account for   0.58%(0.30s) of the runtime)

Apply
------
<% time> <sum %> <apply time> <time per call> <#call> <id> <Apply name>
   0.6%     0.6%       0.328s       1.56e-03s    210   1121   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.6%     1.2%       0.296s       1.41e-03s    210   1202   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.5%     1.8%       0.265s       1.26e-03s    210   1194   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.5%     2.2%       0.250s       1.19e-03s    210   1200   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.5%     2.7%       0.250s       1.19e-03s    210   1120   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     3.2%       0.218s       1.04e-03s    210   1199   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     3.6%       0.218s       1.04e-03s    210   854   GpuDot22(GpuDimShuffle{1,0}.0, GpuElemwise{Mul}[(0, 0)].0)
   0.4%     4.0%       0.203s       9.66e-04s    210   1119   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     4.4%       0.187s       8.91e-04s    210   1195   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     4.7%       0.187s       8.91e-04s    210   1197   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.4%     5.1%       0.187s       8.91e-04s    210   251   GpuDot22(GpuElemwise{Composite{(i0 * (i1 + i2))},no_inplace}.0, <CudaNdarrayType(float32, matrix)>)
   0.3%     5.4%       0.172s       8.17e-04s    210   890   GpuGemm{inplace}(GpuDot22.0, TensorConstant{1.0}, GpuDimShuffle{1,0}.0, GpuElemwise{Composite{(i0 + (-i1))},no_inplace}.0, TensorConstant{1.0})
   0.3%     5.8%       0.172s       8.17e-04s    210   1181   GpuIncSubtensor{InplaceInc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{209460}, Constant{366260})
   0.3%     6.1%       0.172s       8.17e-04s    210   839   GpuDot22(GpuDimShuffle{1,0}.0, GpuElemwise{Mul}[(0, 0)].0)
   0.3%     6.4%       0.172s       8.17e-04s    210   851   GpuDot22(GpuDimShuffle{1,0}.0, GpuElemwise{Mul}[(0, 0)].0)
   0.3%     6.8%       0.172s       8.17e-04s    210   1184   GpuIncSubtensor{InplaceInc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{209460}, Constant{366260})
   0.3%     7.1%       0.172s       8.17e-04s    210   846   GpuDot22(GpuElemwise{Mul}[(0, 0)].0, <CudaNdarrayType(float32, matrix)>)
   0.3%     7.5%       0.172s       8.17e-04s    210   456   GpuDot22(GpuDimShuffle{1,0}.0, GpuElemwise{Mul}[(0, 0)].0)
   0.3%     7.8%       0.156s       7.43e-04s    210   1198   GpuIncSubtensor{Inc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{366260}, Constant{367044})
   0.3%     8.1%       0.156s       7.43e-04s    210   1061   GpuIncSubtensor{InplaceInc;int64:int64:}(GpuIncSubtensor{InplaceInc;int64:int64:}.0, GpuReshape{1}.0, Constant{209460}, Constant{366260})
   ... (remaining 1183 Apply instances account for 91.93%(46.70s) of the runtime)  
```

ReLU Op code:
```
class ReLU(scalar.UnaryScalarOp):
    """
    This is just speed opt. Not for stability.

    """
    @staticmethod
    def st_impl(x):
        return 0.5*(x + abs(x))

    def impl(self, x):
        return ReLU.st_impl(x)

    def grad(self, inp, grads):
        x, = inp
        gz, = grads
        rval = gz * scalar.eq(x, abs(x))
        return [rval]


scalar_relu = ReLU(scalar.upgrade_to_float, name='scalar_relu')
relu = elemwise.Elemwise(scalar_relu, name='relu')

local_relu_to_abs = gof.PatternSub(in_pattern=(relu, 'x'),
                                   out_pattern=(tensor.mul, 0.5, (tensor.add, 'x', (tensor.abs_, 'x'))),
                                   name='local_relu_to_abs',
                                   allow_multiple_clients=True)
opt.register_canonicalize(local_relu_to_abs, 'fast_run', name='local_relu_to_abs')
```

**Comment by f0k** [comment]
I've been thinking about this again -- an actual problem with the current approach is that Theano's automatically derived gradient depends on the input to the rectifier. It should instead be defined in terms of the output, so the forward pass can work in-place. This is possible for the leaky rectifier as well as long as the leakiness is a nonnegative constant.

For reference, the current gradient of y = T.nnet.relu(x) is ((0.5 * topgrad) + (0.5 * topgrad * sgn(x))).
The current gradient of y = T.nnet.relu(x, 0.1) is ((0.55 * topgrad) + (0.45 * topgrad * sgn(x))).

We could define a rectifier placeholder Op that is replaced with a fast forward pass expression and has a `grad` method with a fast backward pass expression that allows in-place operation in both directions.
